src/consensus/aft/raft.h

```diff
@@ -475,22 +481,29 @@ namespace aft
             configurations.back().nodes.end() &&
           conf.find(state->my_node_id) == conf.end())
         {
-          CCF_ASSERT_FMT(
-            !retirement_idx.has_value(),
-            "retirement_idx already set to {}",
-            retirement_idx.value());
-          retirement_idx = idx;
-          LOG_INFO_FMT("Node retiring at {}", idx);
+          become_retired(idx, kv::RetirementPhase::Ordered);
         }
       }
       else
       {
-        if (!new_learners.empty())
+        if (
+          !configurations.empty() &&
+          configurations.back().nodes.find(state->my_node_id) ==
+            configurations.back().nodes.end() &&
+          conf.find(state->my_node_id) == conf.end() &&
+          (is_retiring() ||
+           (is_retired() &&
+            retirement_phase == kv::RetirementPhase::Committed)))
+        {
```

Comment:
I believe this condition is correct, but it's super ugly. I was hoping that it would be the same as for 1tx, but it's not. At the time we transition from `RETIRING` to `RETIRED`, the node is indeed not in `configurations.back()` anymore.

Reply:
That is super ugly... :(

Reply:
I think there may be a better way, if I pass the retiring states instead of the retired ones into `add_configuration`, but it may also create ugliness elsewhere :-/

Reply:
Gave this another try today and it's possibly, but it creates a bunch of complications in `commit`. The core of the problem is that for every actual reconfiguration I get two calls to `add_configuration` as there are two updates to the nodes table and I have to figure out which one's which afterwards. I think we should go ahead with the current version of the ugliness until I work out all of the details for the other one.